arch/x86/core/acpi.c

```diff
@@ -1,103 +1,115 @@
 /*
- * Copyright (c) 2019 Intel Corporation
+ * Copyright (c) 2020 Intel Corporation
  * SPDX-License-Identifier: Apache-2.0
  */
-
 #include <kernel.h>
 #include <arch/x86/acpi.h>
 
-/*
- * Finding and walking the ACPI tables can be time consuming, so we do
- * it once, early, and then cache the "interesting" results for later.
- */
-
-static struct acpi_madt *madt;
+struct acpi_rsdp {
+	char     sig[8];
+	uint8_t  csum;
+	char     oemid[6];
+	uint8_t  rev;
+	uint32_t rsdt_ptr;
+	uint32_t len;
+	uint64_t xsdt_ptr;
+	uint8_t  ext_csum;
+	uint8_t  _rsvd[3];
+} __packed;
+
+struct acpi_rsdt {
+	struct acpi_sdt sdt;
+	uint32_t table_ptrs[];
+} __packed;
+
+struct acpi_xsdt {
+	struct acpi_sdt sdt;
+	uint64_t table_ptrs[];
+} __packed;
+
+static bool check_sum(struct acpi_sdt *t)
+{
+	uint8_t sum = 0, *p = (uint8_t *)t;
 
-/*
- * ACPI structures use a simple checksum, such that
- * summing all the bytes in the structure yields 0.
- */
+	for (int i = 0; i < t->len; i++) {
+		sum += p[i];
+	}
+	return sum == 0;
+}
 
-static bool validate_checksum(void *buf, int len)
+static struct acpi_rsdp *find_rsdp(void)
 {
-	uint8_t *cp = buf;
-	uint8_t checksum = 0;
+	uint64_t magic = 0x2052545020445352; /* == "RSD PTR " */
+
+	/* Physical (real mode!) address 0000:040e stores a (real
+	 * mode!!) segment descriptor pointing to the 1kb Extended
+	 * BIOS Data Area.  Look there first.
+	 */
+	uint64_t *search = (void *)(long)(((int)*(uint16_t *)0x040eL) << 4);
 
-	while (len--) {
-		checksum += *(cp++);
+	for (int i = 0; i < 1024/8; i++) {
+		if (search[i] == magic) {
+			return (void *)&search[i];
+		}
 	}
 
-	return (checksum == 0);
-}
+	/* If it's not there, then look for it in the last 128kb of
+	 * real mode memory.
+	 */
+	search = (uint64_t *)0xe0000;
+	for (int i = 0; i < 128*1024/8; i++) {
+		if (search[i] == magic) {
+			return (void *)&search[i];
+		}
+	}
 
-/*
- * Called very early during initialization to find ACPI tables of interest.
- * First, we find the RDSP, and if found, use that to find the RSDT, which
- * we then use to find the MADT. (This function is long, but easy to follow.)
- */
+	/* Now we're supposed to look in the UEFI system table, which
+	 * is passed as a function argument to the bootloader and long
+	 * forgotten by now...
+	 */
```

Comment:
Obviously in light of #26230 this becomes quite a bit more feasible.  Once that merges this should be the default mechanism for a EFI build, and move up to the top of the list above the expensive choices.